Add MediaSample to MediaSamplesBlock and back conversion

jyavenard opened the pull request:
#### 541443c761723f29fbcbada185237bffa95f5d0a
<pre>
Add MediaSample to MediaSamplesBlock and back conversion
<a href="https://bugs.webkit.org/show_bug.cgi?id=299289">https://bugs.webkit.org/show_bug.cgi?id=299289</a>
<a href="https://rdar.apple.com/161095211">rdar://161095211</a>

Reviewed by Youenn Fablet.

We add the ability to convert a video CMFormatDescription into a VideoTrack.
When converting a MediaSamplesBlock video to a CMSampleBuffer, add handling
of non-displaying sample.
Add {from/to}MediaSample in MediaSamplesBlock to convert respectively from
and to a MediaSampleAVFObjC in order to prepare for serialising the object
over IPC. This allows for C++-only conversion without having to interact
with CMUtilities.

No functional change yet. Tests will be added in a follow-up change which
makes use of those new features.

* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.cpp:
* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.h:
* Source/WebCore/platform/MediaSamplesBlock.cpp: Added.
(WebCore::MediaSamplesBlock::toMediaSample const):
(WebCore::MediaSamplesBlock::fromMediaSample):
* Source/WebCore/platform/MediaSamplesBlock.h:
* Source/WebCore/platform/TrackInfo.h: Add BlockType. When converting from a MediaSample
we record the Box types contained in specific Sample Entries. This guarantees
that even if we encounter a codec not explictly handled (vp8/vp9/avc/hevc)
we will be able to serialise and decode it later without error.
(WebCore::VideoInfo::create):
(WebCore::VideoInfo::VideoInfo):
* Source/WebCore/platform/graphics/cocoa/CMUtilities.h:
* Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
(WebCore::createFormatDescriptionFromTrackInfo):
(WebCore::createVideoInfoFromFormatDescription):
(WebCore::toCMSampleBuffer):
(WebCore::samplesBlockFromCMSampleBuffer):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/300498@main">https://commits.webkit.org/300498@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/27f6e6d58984b2d028f20d58f698cafab38a7302

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/122803 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42515 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33205 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/129429 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/74907 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/038f9f1e-eaa0-449d-bd46-7083934a29cf) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/124679 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43234 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51108 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/93330 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/74907 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/d15f7c68-bfc3-4412-957d-5278612a62dd) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/125755 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/34469 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/109931 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/73971 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/1370b107-3fb6-46ef-aaf3-3f971cca6ed3) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/33450 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28088 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/72921 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/104169 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28301 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/132155 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/49748 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/37871 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/132155 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/50125 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106147 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/132155 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25826 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/25274 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/46519 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49606 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/55358 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/49072 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/52424 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/50755 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/MediaSamplesBlock.cpp:
```diff
@@ -0,0 +1,62 @@
+#else
+    UNUSED_PARAM(sample);
+    UNUSED_PARAM(trackInfo);
+    ASSERT_NOT_REACHED();
```
Why ASSERT_NOT_REACHED here and notImplemented above. Should we align both?

youennf commented on Source/WebCore/platform/TrackInfo.h:
```diff
@@ -120,17 +121,18 @@ struct VideoInfo : public TrackInfo {
     // Used by IPC generator
     friend struct IPC::ArgumentCoder<VideoInfo, void>;
-    static Ref<VideoInfo> create(FourCC codecName, const String& codecString, WebCore::TrackID trackID, FloatSize size, FloatSize displaySize, uint8_t bitDepth, PlatformVideoColorSpace colorSpace, RefPtr<SharedBuffer>&& atomData)
+    static Ref<VideoInfo> create(FourCC codecName, const String& codecString, WebCore::TrackID trackID, FloatSize size, FloatSize displaySize, uint8_t bitDepth, PlatformVideoColorSpace colorSpace, String boxType, RefPtr<SharedBuffer>&& atomData)
```
String&& or const String&.
Ditto below.

youennf commented on Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
```diff
@@ -275,6 +281,53 @@ static CFStringRef convertToCMYCbCRMatrix(PlatformVideoMatrixCoefficients coeffi
+        return nullptr;
+
+    Ref videoInfo = VideoInfo::create();
+    videoInfo->codecName = PAL::CMFormatDescriptionGetMediaSubType(description);
```
This is a bit weird to have VideoInfo being a recounted struct.
It would be also somehow nice to initialise VideoInfo members just once by passing them to the constructor.
This would follow existing patterns and allow to make members const where possible for instance.

jyavenard replied:
VideoInfo is attached to every sample, it's re-used a lot. Don't want to duplicate that content.

jyavenard replied:
And we use the ability to set it to null when serialising. So we don't need to serialise the TrackInfo on every IPC call ; so we send the first MediaSample with a TrackInfo, cache it and all the others sent with the same TrackInfo are sent with null instead.
The serialisation of the TrackInfo/VideoInfo is the biggest task when sending a MediaSamplesBlock

youennf commented on Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
```diff
@@ -275,6 +281,53 @@ static CFStringRef convertToCMYCbCRMatrix(PlatformVideoMatrixCoefficients coeffi
+    Ref videoInfo = VideoInfo::create();
+    videoInfo->codecName = PAL::CMFormatDescriptionGetMediaSubType(description);
+    auto dimensions = PAL::CMVideoFormatDescriptionGetDimensions(description);
+    videoInfo->size = { float(dimensions.width), float(dimensions.height) };
```
```suggestion
    videoInfo->size = IntSize { dimensions.width, dimensions.height };
```

youennf commented on Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
```diff
@@ -275,6 +281,53 @@ static CFStringRef convertToCMYCbCRMatrix(PlatformVideoMatrixCoefficients coeffi
+            RELEASE_LOG_INFO(Media, "kCMFormatDescriptionExtension_SampleDescriptionExtensionAtoms having %ld keys keys expected only 1", extensionCount);
+        else {
+            Vector<const void*> keys(extensionCount);
+            Vector<const void*> values(extensionCount);
```
```suggestion
            Vector<const void*, 1> values(extensionCount);
```
Would this work to remove heap allocation?
Ditto for keys.

jyavenard committed:
Add MediaSample to MediaSamplesBlock and back conversion
https://bugs.webkit.org/show_bug.cgi?id=299289
rdar://161095211

Reviewed by Youenn Fablet.

We add the ability to convert a video CMFormatDescription into a VideoTrack.
When converting a MediaSamplesBlock video to a CMSampleBuffer, add handling
of non-displaying sample.
Add {from/to}MediaSample in MediaSamplesBlock to convert respectively from
and to a MediaSampleAVFObjC in order to prepare for serialising the object
over IPC. This allows for C++-only conversion without having to interact
with CMUtilities.

No functional change yet. Tests will be added in a follow-up change which
makes use of those new features.

* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.cpp:
* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.h:
* Source/WebCore/platform/MediaSamplesBlock.cpp: Added.
(WebCore::MediaSamplesBlock::toMediaSample const):
(WebCore::MediaSamplesBlock::fromMediaSample):
* Source/WebCore/platform/MediaSamplesBlock.h:
* Source/WebCore/platform/TrackInfo.h: Add BlockType. When converting from a MediaSample
we record the Box types contained in specific Sample Entries. This guarantees
that even if we encounter a codec not explictly handled (vp8/vp9/avc/hevc)
we will be able to serialise and decode it later without error.
(WebCore::VideoInfo::create):
(WebCore::VideoInfo::VideoInfo):
* Source/WebCore/platform/graphics/cocoa/CMUtilities.h:
* Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
(WebCore::createFormatDescriptionFromTrackInfo):
(WebCore::createVideoInfoFromFormatDescription):
(WebCore::toCMSampleBuffer):
(WebCore::samplesBlockFromCMSampleBuffer):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/300498@main
